Relax access requests limitations and refactor `AccessRequestsService` state

gzdunek opened the pull request:
As of now, we had two limitations when creating access requests:
1. Resources from root and leaf clusters can't be added to a single request.
2. Role and resource requests can't be mixed.

As it turned out, 1. was an artificial limitation. It's perfectly fine for the backend to have requests from many clusters in a single access request. 
But why did I want to remove it? It's because in the search bar you have access to all resources, so we can't easily block adding resources from a different cluster (and tbh the previous check that was kept in the cluster switcher was really easy to bypass by simply opening multiple access requests tabs). 

However, after I removed this limitation, another problem occurred: we don't store cluster URI for the resources.
To fix this, I had to refactor the access requests service (and its state in particular).
The `{app: {}, node: {}, ...}` was replaced by a `Map` key by resource URI. This ensures that resources are unique among clusters. 
The map value is a request object, which keeps a resource kind, uri, and for servers their hostname.
Previously, the hostname was stored as `kindIds[resourceId] = resourceName || resourceId`. I must say that I didn't like it, mainly   because it was up to the caller to set the name (and it was really easy to miss it). Now it is required by the types for the particular resource type.

As of 2: I moved the check to the access requests service level, because as above, we can't control what the user adds to the request: it may add a role and the go the search bar and add a resource. 
I also changed the state shape, now it's `roles | resources`, not roles & resources` to better reflect the business logic.

TODO: show cluster name in the request checkout of there is more than one cluster in the request.

gzdunek committed:
Remove the check that prevented changing active cluster while making an access request

This was an artificial limitation, we only need to pass the correct cluster name to each requested resource.

gzdunek committed:
Refactor access requests state

The aim of this refactor is to disallow mixing resource and role access requests at the type level.
Another thing is that for resource access requests we need to store cluster name.
Because it's not needed for roles, we have to separate these two resource kinds.

gzdunek committed:
Detect mixing resource and role access request while adding an item instead of when switching resource tab

gzdunek committed:
Rename `getAddedResourceCount` to `getAddedItemsCount` to avoid connections with resource access requests

gzdunek commented on web/packages/teleterm/src/ui/services/workspacesService/accessRequestsService.ts:
```diff
@@ -54,42 +62,249 @@ export class AccessRequestsService {
+    }
+  | { kind: 'role'; roles: Set<string> };
+
+export type ResourceRequest =
```
Why such type? It matches the shape of a resource in the search bar and in the unified resources view, so we will be able to simply call:
```ts
accessRequestsService.addOrRemoveResource(resource);
```

ravicious replied:
I'm not sure if convenience should be the dominant factor when making decisions like this one. I'd focus predominantly on whether it fits the domain and, in this particular case, if we'd ever want to reuse it outside of teleterm.

I think the type is fine though. I feel like the comment you left here on GitHub should be turned into a code comment, so that the next person looking at `ResourceRequest` will know about this.

gzdunek replied:
>I'm not sure if convenience should be the dominant factor when making decisions like this one. I'd focus predominantly on whether it fits the domain and, in this particular case, if we'd ever want to reuse it outside of teleterm.

Ah, I didn't write the whole story. At first I had a type like `{kind: 'server', uri: ServerUri, hostname: string} | {kind: 'database', ...}`. Unfortunately, it required additional logic to map the unified view or search bar resource to this type. Then I thought that I could match the shape of the other resources to make my life easier. After that I happily wrote this comment, saying only about the convenience part :)

I left a JSDoc comment too.

ravicious replied:
The fact that it matches the shape of resources in _both_ the search bar and the unified resources view makes it a much more convincing argument for me than just the fact that it coincidentally matches the shape of resources in the search bar.

ravicious commented on web/packages/teleterm/src/ui/TopBar/Clusters/Clusters.tsx:
```diff
@@ -53,17 +49,7 @@ export function Clusters() {
 
   function selectItem(clusterUri: ClusterUri): void {
     setIsPopoverOpened(false);
-    if (clusters.hasPendingAccessRequest) {
```
It looks like the relevant field can be removed from `useClusters` as well, along with `clearPendingAccessRequest`.

ravicious commented on web/packages/teleterm/src/ui/services/workspacesService/accessRequestsService.ts:
```diff
@@ -54,42 +60,219 @@ export class AccessRequestsService {
-      Object.keys(pendingAccessRequest.windows_desktop).length +
-      Object.keys(pendingAccessRequest.user_group).length
-    );
+    if (pendingAccessRequest.kind === 'role') {
+      return pendingAccessRequest.roles.size;
+    }
+    if (pendingAccessRequest.kind === 'resource') {
+      return pendingAccessRequest.resources.size;
+    }
```
This could also be a switch with a default branch with `satisfies never` to add some extra type safety.

ravicious commented on web/packages/teleterm/src/ui/services/workspacesService/accessRequestsService.ts:
```diff
@@ -54,42 +60,219 @@ export class AccessRequestsService {
+        resources.delete(resource.uri);
+      } else {
+        // Store only properties required by the type.
+        if (kind === 'app' || kind === 'database' || kind === 'kube') {
```
Wouldn't it be better to do it the other way around? As in do `if (kind === 'server')` followed by `else`? Otherwise each new resource type we add will have hostname added.

This too could be a switch but idk if it really needs to be. As long as we have a sensible default that could in theory be applied to each new kind of resource, then I think it's fine to have a simple if else. In this case we do have such a default – each resource must have its own URI.

ravicious reviewed (commented):
I finished midway through the second commit. The change with the types sounds pretty good.

ravicious commented on web/packages/teleterm/src/ui/AccessRequestCheckout/useAccessRequestCheckout.ts:
```diff
@@ -171,12 +198,34 @@ export default function useAccessRequestCheckout() {
     resourceName: string
   ) {
-    workspaceAccessRequest.addOrRemoveResource(kind, resourceId, resourceName);
+    if (kind === 'role') {
+      await workspaceAccessRequest.addOrRemoveRole(resourceId);
+      return;
+    }
+
+    if (
+      kind === 'app' ||
+      kind === 'db' ||
+      kind === 'kube_cluster' ||
+      kind === 'node'
+    ) {
+      await workspaceAccessRequest.addOrRemoveResource(
+        toResourceRequest({
+          kind,
+          resourceId,
+          resourceName,
+          clusterUri,
+        })
+      );
+      return;
+    }
```
Here we're dealing with essentially two groups of enum values. I wonder if it wouldn't be better to drop the second `if`: `workspaceAccessRequest.addOrRemoveResource` already enforces only certain `kind`s through its types.

So if a new kind of resource gets added, it'll be one less place to adjust without sacrificing type safety.

gzdunek replied:
Actually, it turned out that the type `ResourceKind` from `teleterm/useNewRequest` was too wide, it contained kinds `user_group` and `windows_desktop` which we don't support at all.
After I corrected this, I was able to get rid of this `if`, since we narrowed down the type to `'app' | 'db' | 'kube_cluster' | 'node' anyway.

ravicious commented on web/packages/teleterm/src/ui/DocumentAccessRequests/NewRequest/useNewRequest.ts:
```diff
@@ -206,7 +207,29 @@ export default function useNewRequest() {
+      return;
+    }
+
+    if (
+      kind === 'app' ||
+      kind === 'db' ||
+      kind === 'kube_cluster' ||
+      kind === 'node'
+    ) {
```
Same as above.

ravicious commented on web/packages/teleterm/src/ui/services/workspacesService/accessRequestsService.ts:
```diff
@@ -54,42 +60,219 @@ export class AccessRequestsService {
+      };
+    };
+
+type SharedResourceAccessRequestKind = 'app' | 'db' | 'node' | 'kube_cluster';
```
I wasn't sure if it's best to redefine such type here or define it by using a type from the shared package (well, the teleport package in this case). But in the end I guess it doesn't matter much as those are just simple strings and TypeScript will tell us if something doesn't match.

And if we wanted to use a type from the shared package, we'd have to do something like `Exclude<>` or `Extract<>` anyway.

gzdunek replied:
Yeah, I think re-declared type here is easier to read than `Exclude` (I'm also not sure what is the source of truth for this union). 
But we will get a TS error if the types don't match anyway.

ravicious commented on web/packages/teleterm/src/ui/DocumentAccessRequests/NewRequest/useNewRequest.ts:
```diff
@@ -206,7 +207,29 @@ export default function useNewRequest() {
+          kind,
+          resourceId,
+          resourceName,
+          clusterUri,
```
I've looked up how `useNewRequest` gets `clusterUri` and I think the way it does this is incorrect. It gets the URI from `localClusterUri` of the workspace, but each `NewRequest` document is bound to a particular cluster AFAIR.

I haven't been able to confirm this as I temporarily cannot log into my local enterprise cluster due to the MFA changes.

gzdunek replied:
You are right, we should get the cluster URI from the document.

ravicious reviewed (commented):
Made it through the second commit, I'll continue the review tomorrow.

So far everything looks fine, my comments are mostly just tangents about keeping type safety of these changes in the future.

gzdunek committed:
Remove unused fields from `useClusters`

gzdunek committed:
Increase type safety in `getAddedItemsCount`

gzdunek committed:
When storing hostname check if the resource is `server` not the other way around

gzdunek committed:
Remove unnecessary check for resource kind

The original type `ResourceKind` was too wide,
we don't support `user_group` and `windows_desktop` anyway.

gzdunek committed:
Take `clusterUri` from the doc instead of the workspace

gzdunek committed:
Add a comment for `ResourceRequest`

ravicious commented on web/packages/teleterm/src/ui/ModalsHost/modals/ChangeAccessRequestKind/ChangeAccessRequestKind.story.tsx:
Thanks for adding the story. 👍

ravicious commented on web/packages/teleterm/src/ui/ModalsHost/modals/ChangeAccessRequestKind/ChangeAccessRequestKind.tsx:
```diff
@@ -26,51 +24,43 @@ import DialogConfirmation, {
-          onClick={() => onConfirm(toResource)}
-        >
+        <ButtonWarning size="large" block={true} onClick={onConfirm}>
           Confirm
```
```suggestion
          Continue
```

ravicious commented on web/packages/teleterm/src/ui/DocumentAccessRequests/NewRequest/useNewRequest.ts:
```diff
@@ -349,6 +339,8 @@ function getDefaultSort(kind: ResourceKind): SortType {
 
-export type ResourceKind = ResourceIdKind | 'role';
+export type ResourceKind =
+  | Exclude<ResourceIdKind, 'user_group' | 'windows_desktop'>
```
I was going to say that this is technically less future-proof, because with every added kind, TypeScript will automatically assume that Connect supports it. So it could be better to explicitly say which ones we support instead of which ones are not supported.

OTOH if there's a place in the code that does an exhaustive check on this type, then adding a new kind will automatically force people to either implement support in Connect or explicitly add it to the ignore list here.

On the other other hand, I don't think there's such place with an exhaustive check in the project at the moment.

But overall it should be fine to leave it like that at the moment.

---

Update: It's not really true though, I just tried adding a new kind. There's a bunch of errors because there are other places in the code which explicitly list which kinds Connect supports.

gzdunek replied:
Actually, I like the version with explicit kinds more :)
Changed.

ravicious commented on web/packages/teleterm/src/ui/TopBar/AdditionalActions.tsx:
```diff
@@ -106,7 +106,7 @@ function useMenuItems(): MenuItem[] {
       onNavigate: () => {
         const doc = documentsService.createAccessRequestDocument({
-          clusterUri: activeRootCluster.uri,
+          clusterUri: workspacesService.getActiveWorkspace().localClusterUri,
```
Hmm, this makes me think that we actually shouldn't touch this… It used to accept only root cluster URIs but now it accepts local cluster URIs too?

From what I see, I guess picking the correct cluster in this doc just never worked correctly. The doc used to be always associated with a root cluster, but when adding a new request we'd always look at which cluster is currently selected.

I'd just revert this change, as properly fixing this is probably outside of the scope of this PR.

gzdunek replied:
Yeah, I will revert this. 
We may end up not needing this change at all, if we remove resource requests from `NewRequest`. We will only have role requests there (which are always done against the root cluster).

ravicious reviewed (approved):
🚢

gzdunek committed:
Confirm -> Continue

gzdunek committed:
Make supported kinds explicit

gzdunek committed:
Revert "Take `clusterUri` from the doc instead of the workspace"

This reverts commit b5a93cd611c55742e80515d404becb07319818e4.

gzdunek committed:
Merge branch 'refs/heads/master' into gzdunek/relax-access-requests-limitations

gzdunek committed:
Adjust new `useAccessRequestCheckout` test
